src/YearnFed.sol

```diff
@@ -0,0 +1,162 @@
+        //as that supplies greatest possible amount into vault
+        /*
+        if( amount > _maxDeposit()){
+            amount = type(uint256).max;
+        }
+        */
+        require(amount <= _maxDeposit(), "AMOUNT TOO BIG"); // can't deploy more than max
+        underlying.mint(address(this), amount);
+        uint shares = vault.deposit(amount, address(this));
+        require(shares == 0, 'Supplying failed'); //Probably an unnecessary require
+        supply = supply + amount;
+        emit Expansion(amount);
+    }
+
+    /**
+    @notice Withdraws an amount of underlying token to be burnt, contracting DOLA supply
+    
+    @dev Its recommended to always broadcast withdrawl transactions(contraction & takeProfits)
+    through a frontrun protected RPC like Flashbots RPC.
+    
+    @param amount The amount of underlying tokens to withdraw. Note that more tokens may
+    be withdrawn than requested, as price is calculated by debts to strategies, but strategies
+    may have outperformed price of underlying token.
+    */
+    function contraction(uint amount) public {
+        require(msg.sender == chair, "ONLY CHAIR");
+        uint underlyingWithdrawn = _withdrawAmountUnderlying(amount, maxLossBpContraction);
+        require(underlyingWithdrawn <= supply, "AMOUNT TOO BIG"); // can't burn profits
+        require(underlyingWithdrawn > 0, "NOTHING WITHDRAWN");
+        underlying.burn(underlyingWithdrawn);
+        supply = supply - amount;
+        emit Contraction(underlyingWithdrawn);
+    }
+
+    /**
+    @notice Withdraws the profit generated by yEarn vault
+
+    @dev See dev note on Contraction method
+    */
+    function takeProfit() public {
+        uint expectedBalance = vault.balanceOf(address(this))*vault.pricePerShare()/10**vault.decimals();
+        if(expectedBalance > supply){
+            uint expectedProfit = expectedBalance - supply;
+            if(expectedProfit > 0) {
+                uint actualProfit = _withdrawAmountUnderlying(expectedProfit, maxLossBpTakeProfit);
+                require(actualProfit > 0, "NO PROFIT");
+                underlying.transfer(gov, actualProfit);
+            }
+        }
+    }
+
+    /**
+    @notice calculates the amount of shares needed for withdrawing amount of underlying, and withdraws that amount.
+
+    @dev See dev note on Contraction method
+
+    @param amount The amount of underlying tokens to withdraw.
+    @param maxLossBp The maximally acceptable loss in basis points. 1 = 0.001%
+    */
+    function _withdrawAmountUnderlying(uint amount, uint maxLossBp) internal returns (uint){
+        uint sharesNeeded = amount*10**vault.decimals()/vault.pricePerShare();
+        return vault.withdraw(sharesNeeded, address(this), maxLossBp);
+    }
+
+    /**
+    @notice calculates the maximum possible deposit for the yearn vault
+    */
+    function _maxDeposit() view internal returns (uint) {
+        return vault.depositLimit() - vault.totalAssets();
```

Comment:
this can revert.
here is a solution that will not revert...
```
        uint depositLimit = vault.depositLimit();
        uint totalAssets = vault.totalAssets();
        if depositLimit > totalAssets {
            return depositLimit - totalAssets;
        }
        return 0;
```

Reply:
Thank you